core/src/main/java/io/grpc/internal/ClientCallImpl.java

```diff
@@ -466,22 +466,25 @@ public final void runInContext() {
     }
 
     @Override
-    public void messageRead(final InputStream message) {
+    public void messagesAvailable(final MessageProducer producer) {
       class MessageRead extends ContextRunnable {
         MessageRead() {
           super(context);
         }
 
         @Override
         public final void runInContext() {
+          InputStream message;
           try {
-            if (closed) {
-              return;
-            }
-            try {
-              observer.onMessage(method.parseResponse(message));
-            } finally {
-              message.close();
+            while ((message = producer.next()) != null) {
+              if (closed) {
+                return;
+              }
+              try {
+                observer.onMessage(method.parseResponse(message));
+              } finally {
```

Comment:
Fixed. We spoke offline and the problem would occur with exceptions in handling the message when the producer potentially (as in the subsequent parts of this PR) has additional queued messages.